internal/expr/parser.go

```diff
@@ -17,17 +18,33 @@ type queryPart interface {
 	part()
 }
 
-// valueAccessor stores information for accessing a Go value. It consists of a
-// type name and some value within it to be accessed. For example: a field of a
-// struct, or a key of a map.
-type valueAccessor struct {
+// valueAccessor stores information about how to access a Go value. For example,
+// by index, by key, or by slice syntax.
+type valueAccessor interface {
+	TypeName() string
```

Comment:
Should this be `typeName`? i.e. not exported?

Reply:
The benefit of having it in upper case is that it does not clash with the field name `typeName` of `memberAccessor`, and following the same naming convention as `queryPart` in the same file which has the `String` method.

Reply:
The `String` method is public so that it implements the `Stringer` interface https://go.dev/tour/methods/17. That type is used by the `fmt` package when printing things.
To me it feels wrong to use a public method here, it implies that there is a reason for it. You could change the field name in `memberAccessor` to `outerType` maybe? As it is in `typeMember`. Its a shame they decided not to allow interfaces to be implemented by values in a struct in Go, it would help a lot here.

Reply:
After all the discussions we had on how to name the accessors, I feel that they are now looking good and I don't want to change it back to `outerType`. I have a few ideas, what do you think of the following?
1. Using `getTypeName` for the interface. Even if it is discouraged in general, it would make sense in this case and there are examples of using the `get` prefix for name clashes like protobuf.
2. Changing the field name to `outerTypeName`.
3. Changing the field name to `typName`.

Reply:
Using `getTypeName` makes the most sense here imo.